internal/guest/runtime/hcsv2/container.go

```diff
@@ -181,3 +185,20 @@ func (c *Container) GetStats(ctx context.Context) (*v1.Metrics, error) {
 func (c *Container) modifyContainerConstraints(ctx context.Context, rt prot.ModifyRequestType, cc *prot.ContainerConstraintsV2) (err error) {
 	return c.Update(ctx, cc.Linux)
 }
+
+// adjustExecSpec adjusts an OCI runtime specs process field provided for an execed process to contain some of the settings set on the containers spec.
+// Some of the OCI spec settings are configured in the guest as there's not enough information available on the host to set them, so the spec set for the
+// exec may be missing some things that were present in the final version of the containers spec.
+func adjustExecSpec(containerSpec *oci.Process, execSpec *oci.Process) {
```

Comment:
Instead of doing this, can we just set the user on the runtime spec in cri, and wait to check if the user actually exists until the container is being started by the gcs?

Reply:
I suppose we could actually check if the snapshotter is set to windows-lcow and skip the bit where we check the rootfs if the user exists https://github.com/containerd/containerd/blob/main/oci/spec_opts.go#L555-L593

Reply:
For upstream it's worth thinking about what's the "right way" to solve this. Checking the snapshotter seems like more of a hack, since really how we prepare images should not be directly related to the fact that the Linux host is in a VM.

Reply:
Kata might run into the same issue, I'm not sure they handle this, would be worth looking. Possibly if the GOOS is Windows might be a better tip off than the snapshotter.

Reply:
If I'm reading their architecture docs correctly looks like they share the rootfs prepared on the host to the guest via 9p (in the default config) so they wouldn't actually hit our issues here.

Reply:
@kevpar Let me know if the GOOS method seems like a sound solution to you, and I'll get to work.

Reply:
I think that is a decent approach, though I wonder if there is a way to get a more explicit signal. Perhaps switching behavior based on the runtime being used?

Reply:
Aha, the[ `containers.Container`](https://github.com/containerd/containerd/blob/7ddf5e52ba738e868b70807797c79c8e54da3497/containers/containers.go#L30) type has a runtime field that might carry what we need for that suggestion. Let me investigate, that seems like a good idea.

Reply:
Aaand it's all comin back to me now. So there **is** something we can do, although it's.. odd for some cases and I'd like your feedback. Containerd accepts the following pairs as a string for users for a container:

1. username 
2. uid
3. username:group
4. uid:gid
5. uid:group
6. username:gid

All of these regardless of which combination get translated to a uid:gid pair that is set on the spec (by this inspecting the rootfs process they have currently). If the user passes a uid:gid string pair (case number 4) and it successfully parses as such, [they just assign it to the spec and do no further validation](https://github.com/containerd/containerd/blob/main/oci/spec_opts.go#L539-L554), this is the happy path. The tricky bit is if they pass any of the other combinations. They all pretty much lead to needing to read from disk to verify/grab the actual uid:gid pairing for the string provided. We obviously can't do that and we really just need to defer this as we've established. We'd ideally like a place to put the user string as provided so we can do the exact same process in the guest, while having something to check against on the User field to see if someone requested a custom user.

I think we might be able to use [the username field on the spec](https://github.com/opencontainers/runtime-spec/blob/0d6cc581aeeaacee38ce707ae651a0e122f76852/specs-go/config.go#L99) to get around the other cases. The odd bit is that this field has the platform for the field defined as only for Windows (and I guess in technical terms we are using it on Windows 😋, at least temporarily). What do you think about storing the user string provided in this field and then doing the checks in the guest as we do now (we can also get rid of our annotation approach with this afterwards)? This solves the issue of not knowing if they provided an explicit uid 0 and gid 0 or the spec simply just has default values as we can check if the username field is an empty string to tell. @ambarve @kevpar, what do you think?

Reply:
Btw, we do have some custom logic in some of the oci options methods in containerd to check if we're running LCOW, they check if the Windows field is filled in on the runtime spec. This one doesn't check if Linux is filled in, but the method itself makes no sense for Windows, should probably but a block on that in case someone tried to use it for Windows containers. https://github.com/containerd/containerd/blob/main/oci/spec_opts.go#L719-L722